### PR TITLE
Make Overhang Printable

### PR DESCRIPTION
**Pull request by odie5533**
Adds feature from issue #1534 

Based on Cura's ConicalOverhang algorithm by @bagelorb: https://github.com/Ultimaker/CuraEngine/blob/main/src/ConicalOverhang.cpp

Algorithm implementation by @Noisyfox. I added the settings options.

Sample file: [Overhang Demo.stl.zip](https://github.com/SoftFever/OrcaSlicer/files/12073932/Overhang.Demo.stl.zip)

Other sample files are in the issue #1534 (Knob and Yoda).

Disabled

orca-slicer_K9PN2e16pb

Enabled

orca-slicer_hugoOnjPCM
orca-slicer_P5edHNbojg

### CODE REVIEW

**Comment by SoftFever** [comment]
Cool! Thank you all.
Running CICD builds so that others can try it out and test it.

**Comment by uberjay** [comment]
> Cool! Thank you all. Running CICD builds so that others can try it out and test it.

How might I download the CI assets? I guessed at a handful of different URL schemes, but failed to figure it out. Perhaps I'm missing something obvious. 😳

**Comment by odie5533** [comment]
> > Cool! Thank you all. Running CICD builds so that others can try it out and test it.
> 
> How might I download the CI assets? I guessed at a handful of different URL schemes, but failed to figure it out. Perhaps I'm missing something obvious. 😳

https://github.com/SoftFever/OrcaSlicer/actions?query=branch%3Aoverhang-settings

If you click one of the builds like https://github.com/SoftFever/OrcaSlicer/actions/runs/5581868427, it will have Artifacts at the bottom.

File: src/libslic3r/PrintObjectSlice.cpp [hunk 1]
```diff
@@ -1166,6 +1170,93 @@ void PrintObject::slice_volumes()
...
+
+        for (size_t region_id = 0; region_id < this->num_printing_regions(); ++region_id) {
+            // export_to_svg(debug_out_path("Surface-obj-%d-layer-%d-region-%d.svg", id().id, layer->id(), region_id).c_str(),
+            //               layer->m_regions[region_id]->slices.surfaces);
+
+            // Calculate the scaled upper poly that belongs to current region
+            auto p = intersection_ex(upper_layer->m_regions[region_id]->slices.surfaces, upper_poly);
+            // And now union it
+            ExPolygons layer_polygons = to_expolygons(layer->m_regions[region_id]->slices.surfaces);
+            layer->m_regions[region_id]->slices.set(union_ex(layer_polygons, p), stInternal);
```
Comment (SoftFever) [inline src/libslic3r/PrintObjectSlice.cpp:1254]: We might need to merge the polygons per type.
Not sure if here is the reason why the issue happened(ref to me comment in the PR thread)

**Comment by Noisyfox** [comment]
An I believe the issue is I did not update the Layer::lslices after altering the surfaces.

**Comment by Noisyfox** [comment]
Ah I just got an idea. We should apply the conical overhang after the original slicing and BEFORE any size compensation (elephant foot/hole/contour compensations, etc.)

**Comment by SoftFever** [comment]
> Ah I just got an idea. We should apply the conical overhang after the original slicing and BEFORE any size compensation (elephant foot/hole/contour compensations, etc.)

_oops, I tagged the wrong person(the original Cura author) in my previous comment._

Sounds promising.

File: src/libslic3r/PrintObjectSlice.cpp (hunk 1 above)
Comment (Noisyfox) [reply src/libslic3r/PrintObjectSlice.cpp:1254]: There is nothing to do with the layer type as at this moment all surfaces are `stInternal` until `detect_surfaces_type()` is called later.

**Comment by Noisyfox** [comment]
I've done the change here in my own repo https://github.com/Noisyfox/OrcaSlicer/pull/1. Couldn't update this PR though.

**Comment by SoftFever** [comment]
> I've done the change here in my own repo [Noisyfox#1](https://github.com/Noisyfox/OrcaSlicer/pull/1). Couldn't update this PR though. Could you help pulling those changes to here @SoftFever thx.

It's probably better for you to create the PR in this case as we might have future changes?
@odie5533 Let me know if you are fine with it. 
Both of you will be the co-authors

**Comment by odie5533** [comment]
I agree, @Noisyfox should make the PR to manage it. I jumped the gun a bit to get it in the CI pipeline.

File: src/libslic3r/PrintObjectSlice.cpp (hunk 1 above)
Comment (SoftFever) [reply src/libslic3r/PrintObjectSlice.cpp:1254]: Make sense

**Pull request closed without merging**